### ISSUE TITLE
Adding New hyperparameter optimization tools

### ISSUE THREAD

**Issue by ianpjohnson** [deeplearning4j/Arbiter#45]
The current Arbiter has a couple of starter hyperparameter  strategies: Grid & Random Search (and apparently a closed Spearmint) . It would be nice to enhance these with existing "best of breed" tools out there (e.g. Spearmint. SMAC, TPE, BayesOpt etc and/or derivates)

One problem is that many of these are implemented in Python/C++ or other languages, so it would be good if we had a way to allow their use without necessarily porting them to Java. Usually it is possible to invoke the DL4J runs from the command line but it is clumsy and non trivial to integrate with the existing DL4J optimization  mechanisms. If external tools can be integrated with limited engineering effort they can potential be part of future Arbiter mechanisms perhaps including adversarial, parallel etc where we might run several  hyperparameter tools concurrently and/or in competition

To this end i am proposing to "bridge" into the existing hyperparameter  mechanism by implementing "external" strategies. They might be connected, for example, by sockets or other mechanisms.

A few things this bridging mechanism might be able to do would be:
- To export the hyperparameter  ranges/types/constraints to the external tool - each has its own format for defining parameters and it would be nice to keep these in 1 place (Arbiter or another common format) and translate automatically for each tool preferably on the fly
- To pool the experiment results and allow each tool to see the results from other tools to save them repeating the same or similar experiments, or to add existing results to their own space for further analysis

**Comment by ianpjohnson** [comment]
It might also be interesting to tracking multiple minima, not just the "best" - Might be interesting to look at meta learning - https://www.aaai.org/ocs/index.php/AAAI/AAAI15/paper/viewFile/10029/9349

If anyone knows of, or bumps into some stuff to track potential "hot spots", let me know - else can just knock up something rudimentary. Whether this is abstractable across tools i don't know (I doubt it at this time) so something in Arbiter might be useful

Am not looking to innovate too much here, but limited adoption of some of the later stuff out there must be worthwhile, at the "right price"

**Comment by ianpjohnson** [comment]
A simple "baseline" for sensible ranges of each parameter would be a great idea. I know this would likely vary per model but we have to start somewhere. Some hyperparameter  optimization tools support conditional variables which can greatly reduce the search cost, although these are not universally available or advised (Spearmint ??) . Some values would be just lunatic and its not always clear what is available so the baseline list would help, perhaps without engineering effort

**Comment by saudet** [comment]
If there are good C++ libraries for that, it's something with could do with JavaCPP:
https://github.com/bytedeco/javacpp-presets/wiki/Create-New-Presets

**Comment by ianpjohnson** [comment]
Possibly - BayesOpt comes to mind http://rmcantin.bitbucket.org/html/ (this is not a recommendation - i haven't tried it YET) and would, of course be a stronger engineering solution (as per DL4J remit) but it might add more difficulties as well (deployment, licensing as well as technical) - perhaps initially a "bridged" solution, as long as it has low engineering cost, and then an embedded solution if it proves to add value) I know DL4J is not so much intended for "playing" but most of the tools out there are python and a bridge could be a halfway house

**Comment by ianpjohnson** [comment]
At the moment i have an ExternalSearchCandidateGenerator which listens on a socket for requests of 2 types -
-parameters sends a list of parameters (say 
  x0=Continuous(min=1.0E-4,max=0.1)
  x1=Integer(min=16,max=512)
and a util class to convert to JSON for Spearmint
{
  "language" : "PYTHON",
  "main-file" : "socket_client.py",
  "experiment-name" : "",
  "likelihood" : "NOISELESS",
  "variables" : {
    "x0" : {
      "type" : "float",
      "size" : 1,
      "min" : "1.0E-4",
      "max" : "0.1"
    },
    "x1" : {
      "type" : "int",
      "size" : 1,
      "min" : "16",
      "max" : "512"
    }
  }
}

and a socket_client.py client that relays jobs to the ExternalSearchCandidateGenerator  and replies back to Spearmint - simples 

Adding another config generator for Smac and fiddling with the socket client (it has one but it either has to be modified or a dfrent SmacSearchCandidateGenerator  assembled) and it should work similarly - an initial connect to generate config then run to generate/process candidates

**Comment by ianpjohnson** [comment]
Each tool has its own specialty it seems e.g dynamic stopping criteria etc - the protocol between the tool and the CandidateGenerator in DL4J could be enhanced to offer control over the learning process as it progresses. Seems a lot of work going on here with early stopping, freeze thaw, time limits etc

**Comment by ianpjohnson** [comment]
I don't have a firm view with parallelism though. Many of the tools are synchronous so offer limited parallelism, others have it in varying degrees - a socket allows a limited scaleout from the tool but it's better i think to delegate to some DL4J hosted mechanism for consistency rather than causing confusion with tool side mechanisms

**Comment by agibsonccc** [comment]
Yeah the license on bayesopt is  a no go ;/

**Comment by ianpjohnson** [comment]
Further to my comments on the earlyadopters channel:
Am looking at plugging a "concrete candidate" into the candidate generation framework - from the CandidateSearch aspect, which seems a semantically appropriate place. I see a graph of Parameters in a space - each is a mapping/generation mechanism. I would like to pass a concrete value from an external optimisation tool without any distribution  mapping as the external tool might already be applying distribution mapping 

Making each an FixedValue would be fine (with mutator) but is then not suited to switch in alternate search methods and i no longer have the meta data required to setup the external tool

My thought was simply to add a setValue to each leaf ParameterSpace type which, when used, overrides the getValue implementation.  This avoid issues with discretization and multiple distribution mappings (though my jury is still out of retaining DL4J distribution mapping in some cases)

**Comment by ianpjohnson** [comment]
If the min, max on ContinuousParameterSpace was exposed (As you do in IntegerParameterSpace - it is in toString only) then i could map into normalized space, allowing a choice of value distribution in DL4J or not.  That would avoid any need to override value and be in the spirit of the existing implementation 

The only other alternative i can think of is to express parameter ranges in the external tools as normalized which hardly aids in transparency and makes explicit conditional parameters a no-go area

DiscreteParameterSpace is ok (I think) as i have numValues

So getMin, getMax on ContinuousParameterSpace  and getName on leaf ParameterSpace is enough for me i think.

**Comment by ianpjohnson** [comment]
Random Search
![image](https://cloud.githubusercontent.com/assets/19364563/19623992/9ff69f14-98da-11e6-8150-4deb4b980f19.png)

Spearmint 
![image](https://cloud.githubusercontent.com/assets/19364563/19623996/ba42ee40-98da-11e6-84ab-245001a4300e.png)

SMAC next ....

**Comment by ianpjohnson** [comment]
Spearmint uses an embedded python fragment to connect to a simple generic socket server in the ExternalSearchCandidateGenerator which listens to  the learning progress and relays a response back.

SMAC has an existing TCP experiment runner but it sends a java serialized object - to avoid the need for a custom DL4J Candidate Generator i chose instead to get SMAC to invoke a command line Java socket client which re-formats the arguments to suit the generic ExternalSearchCandidateGenerator  and translates the response back for SMAC

There is also a small mapping to convert a generic Parameterspace  defifnition in DL4J to a SMAC .pcs file (or a Spearmint config.json mapping)

**Comment by ianpjohnson** [comment]
SMAC - It was, at least, more interesting to see it was doing something different
![image](https://cloud.githubusercontent.com/assets/19364563/19630669/e201c044-9986-11e6-93cf-78be1ab5f442.png)

**Comment by ianpjohnson** [comment]
BayesOpt/TPE next - whichever gives way first (maybe Yelp/MOE) or anything else that jumps up at me - any suggestions ?

**Comment by ianpjohnson** [comment]
**BayesOpt** - http://rmcantin.bitbucket.org/html/index.html

C++ got it working thru Java Socket client for now but looks easy enough to socket straight from C++
Only have it working with N dimensions normalized for now so just wrap as a candidate and ..
![image](https://cloud.githubusercontent.com/assets/19364563/19632839/b4013b4a-99a2-11e6-949c-57e29899aa93.png)

**Comment by agibsonccc** [comment]
Could we see some pull requests for this stuff?

**Comment by ianpjohnson** [comment]
At the moment I am lashing together a mix of utilities, a Candidate Generator, the odd bit of python/C++ - when it stabilizes i'll try and work out how to assemble it into a cohesive entity . It is really a bitty "bridge kit" that you can draw from rather than a cohesive piece of code. But YES i will eventually send a PR request. At this moment it doesn't require any DL4J changes but the suggestions i made earlier (min/max for ContinuousParameterSpace  and name on leaf parameter spaces) will make some of them a little more friendly to use. There are choices for each tool and it is difficult to finalize until i have passed through the common tools out there - just a few to go, perhaps later this week (TPE/MOE perhaps)

At the moment it is (at least theoretically) possible to run multiple tools at the same time. I will add the option to distribute results from each tool to all other contributing tools - they won't use it to start with but hey - they will have the results on offer if they can.

I also haven't even started to properly address automating the config space generation for the tools (just converter fragments) or addressing parallel options within each tool  vs the parallel options already within Arbiter

I would like to address learning "control" for "smart" early stopping options and perhaps even (though the bar is high) influence over model serialization to allow some form of Freeze/Thaw style of training

At that point it might be useful enough to use but this "blog" is really a heads up on where i am aiming - and looking for feedback. It is really too much of a mess to PR right now - lol

**Comment by ianpjohnson** [comment]
I am intrigued as to whether you have interest or ideas on how multiple running tools might be co-ordinated towards a common goal rather than just running them as an .. incoherent horde. And how/where to introduce points in Arbiter where such behavior can be hosted.

Do you have plans ?

**Comment by ianpjohnson** [comment]
BayesOpt is a frustration - it only support Continuous (real) or Discrete - not a mix- so a mix of Continuous & few values Discrete just don't mix well. Looks like its just best to use Discrete  and apply fine grained discretisation - or maybe an option

Shame coz its damned fast - in damned fast mode - have interested in it for my own stuff

**Comment by ianpjohnson** [comment]
BayesOpt doesn't work well with Discrete parameters. It has 2 kinds, specified and discretised. Discretised do not work well at all out of the box - it keeps getting stuck well out of minima. It is recognised as an issue that it's not best with Discrete  parameters and there are suggestions for dfrent kernels but it doesn't bode well. For now it good with all continuous - a bit disappointing, but i might return another day.
But lets not get bogged down - i'll try and look at more later in the week

**Comment by AlexDBlack** [comment]
Apologies for not replying sooner.
OK, so goal here is to bridge to these existing tools (Spearmint, SMAC, BayesOpt etc). Sounds good in principle.


One approach here: might be better to implement with HPOlib instead of each library individually?



**Licensing**
Most of the licences aren't great for us...
Spearmint: custom "Academic and Non-Commercial Research Use Licence"
SMAC: AGPLv3 licence - requires any modifications to be made open source http://www.gnu.org/licenses/agpl-3.0.html
BayesOpt: AGPLv3 also
HPOlib: GPL (but relies on spearmint/SMAC/hyperopt)
Hyperopt: custom but fairly permissive licence?

So we'd have to keep all of this in a separate repo (definitely can't go in main Arbiter repo - MOE possibly being an exception?).

**Conditional Variables**
Not currently supported in Arbiter. Doesn't matter for random search. Gaussian Processes (i.e., spearmint) don't really support them either if memory serves me.

**Baseline / sensible parameter ranges**
Hm... not sure to what extent this is possible. Obviously some values are bad - learning rate = 1000 for example - but hard to generalize in a useful way. Users that understand enough about neural networks shouldn't need this; for users that don't, any set of baseline parameter ranges might not sufficiently constrain the search space to be useful...
Open to suggestions on how this could be implemented, however.


**plugging a "concrete candidate" into the candidate generation framework**
The way Arbiter is set up: candidates are generated as a double[] (each value is 0 to 1), which gets translated to the appropriate real-valued, discrete or integer values.
In the late of discrete/integer, we just use ranges.

So in translating the candidates from some other tool, it might be sensible to map whatever it generates back to the 0-1 range, and pass that to the ParameterSpace to do the translation?

**"If the min, max on ContinuousParameterSpace was exposed..."**
We can (and should) add a getter for the min/max values here. And a getter list of values for DiscreteParameterSpace. No reason not to have this. Feel free to send us a PR for that.


**MOE**
This is another gaussian process tool. Apache 2.0 licenced though - so an integration with this could be added to the main repo. I've never used it, however.


**how multiple running tools might be co-ordinated towards a common goal**
Assuming the hyperparameter configuration spaces overlap, it should be possible in principle to report results generated from one tool, to any of the other tools. Might be possible to somehow weight candidates based on how good they are according to multiple tools (bit like a ML ensemble).
Beyond that: haven't thought about it. That'd be something to consider some way down the line.

**Comment by ianpjohnson** [comment]
* Looked at HPOlib but wanted to avoid the lowest common denominator affect
* I agree on licenses and seperate repo - was thinking of making it easy to integrate both current and future tool implementations as easily as possible - leave licensing issues up to the individual user - you have a closed source Spearmint - is offering open an source AC tool contrary to that ? I wouldn't expect any of these tools in the Arbiter repo - but the common bridging mechanism, yes. I recognise that Skymind might be more interested in integrating mature, active solutions rather than stale projects but this is supposedly a developing field and it must be nice to "tick the box" at least
* plugging a "concrete candidate" - no longer desired - will map into normalised space - then we have distribution mapping and a common range of values across multiple tools cncurrently
* min/max & name - ill do that soon
* Yelp/MOE - high on my list - Dockerised/REST when i first looked - If the license suits DL4J better ill look at it ahead of TPE - am interested in looking at JavaCPP so don't mind taking a shot at JavaCPPify it
* Spearmint is known not to work well with conditional parameters - SMAC is not so bad but i have not tested - conditional parameters support in tools is very patchy but i think that will likely improve over time - i guess its possible to add by hand for a while until its clear its useful (so i will make sure that tool config can be auto and/or part auto)
* Hyperopt is TPE (i think)

Just keep swimming ..
